skfda/preprocessing/dim_reduction/projection/_fpca.py

```diff
@@ -1,33 +1,425 @@
-"""Functional principal component analysis.
-"""
+"""Functional Principal Component Analysis Module."""
 
 import numpy as np
+import skfda
+from abc import ABC, abstractmethod
+from skfda.representation.basis import FDataBasis
+from skfda.representation.grid import FDataGrid
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.decomposition import PCA
 
-from ....exploratory.stats import mean
+__author__ = "Yujian Hong"
+__email__ = "yujian.hong@estudiante.uam.es"
 
 
-def fpca(fdatagrid, n=2):
-    """Compute Functional Principal Components Analysis.
+class FPCA(ABC, BaseEstimator, TransformerMixin):
+    """Defines the common structure shared between classes that do functional
+    principal component analysis
 
-    Performs Functional Principal Components Analysis to reduce
-    dimensionality and obtain the principal modes of variation for a
-    functional data object.
+    Attributes:
+        n_components (int): number of principal components to obtain from
+            functional principal component analysis. Defaults to 3.
+        centering (bool): if True then calculate the mean of the functional data
+            object and center the data first
+        components (FDataGrid or FDataBasis): this contains the principal
+            components either in a basis form or discretized form
+        component_values (array_like): this contains the values (eigenvalues)
+            associated with the principal components
+        pca (sklearn.decomposition.PCA): object for principal component analysis.
+            In both cases (discretized FPCA and basis FPCA) the problem can be
+            reduced to a regular PCA problem and use the framework provided by
+            sklearn to continue.
+    """
+
+    def     __init__(self, n_components=3, centering=True):
+        """FPCA constructor
+
+        Args:
+            n_components (int): number of principal components to obtain from
+                functional principal component analysis
+            centering (bool): if True then calculate the mean of the functional
+                data object and center the data first. Defaults to True
+        """
+        self.n_components = n_components
+        self.centering = centering
+        self.components = None
+        self.component_values = None
+        self.pca = PCA(n_components=self.n_components)
```

Comment:
The scikit-learn guidelines recommend that the __init__ method should be "dumb", and the checks and transformations of the arguments should be done in the `fit` method. Also attributes set in `fit` should end in a trailing underscore.

Reply:
Fixed. 
However, when passing the check function for estimators, it treats X as numpy ndarray and does not accept that I am using attributes from FDataBasis. Is there a fix to that?

Reply:
Do not check using the functions from scikit-learn. Maybe we should have our own checking functions for estimators, but for now do not perform any check.